### ISSUE TITLE
ADSR output stuck at NaN sometimes

### ISSUE THREAD

**Issue by FrankBuss** [VCVRack/Rack#1393]
With this patch:
[nan.vcv.zip](https://github.com/VCVRack/Fundamental/files/3317838/nan.vcv.zip)
sometimes one output is stuck at NaN:
![nan](https://user-images.githubusercontent.com/1707296/59976284-6871e700-95c2-11e9-8351-be7e19e8419f.png)
Note: you can't use the currently released Specific Value module, because it crashes with NaN input, I submitted already another bug report for the author. But you can see the stuck channels as well when the output LED is always turned on.
Sometimes it is 2 channels, sometimes no channels, most of the time 1 channel. Looks like always a channel from 5 to 8, didn't see it happen on channel 1 to 4. Tested on Linux Debian. Happens with other gate sources as well, like from the Clocked module (with an additional Merge), or the PolyGenerator module.

**Comment by AndrewBelt** [comment]
Fixed in 01d60a0. Review commit and check if this solves your issue.

**Comment by FrankBuss** [comment]
Thanks, this solves the problem.
BTW, there are some warning when I compile Rack on my Linux system. I didn't check it, but one says "array subscript is above array bounds": https://pastebin.com/7d4h8BFc

**Comment by AndrewBelt** [comment]
I've seen that before on old compilers but have no idea it's talking about. Newer compilers don't have the message so probably a compiler bug.

**Comment by FrankBuss** [comment]
Ok. And probably a good idea to change ADSR.cpp to "float_4 attacking[4];" as well, because it is misleading and unnecessary currently.

**Comment by FrankBuss** [comment]
BTW: the union assumes `sizeof(float)==4`. Might be true for all platforms, and maybe you check this already elsewhere, but in the stackoverflow comment someone mentioned that you could write `static_asset(sizeof(float) == 4, "ERROR")`, which is evaluated at compile time.

I might not like C++, but always amazing how many patches and quirks the language designers add on top of each other. I wish everyone would use Common Lisp, where you can run any code at compile time without cryptic syntax or the art of template meta-programming, and it has much better and easier lambda support, all since the beginning :-)

**Comment by AndrewBelt** [comment]
>BTW: the union assumes sizeof(float)==4.

Eh who cares

Okay, I've found a better solution. We can set a default constructor `Vector() = default;` which will make `Vector` behave like a POD type just like `float`, so that `float_4 x[4] = {};` initializes to 0 and `float_4 x[4];` remained undefined.

Fixed in 9f9e85d. Note that `float_4 attacking[4] = {};` is required now.

**Comment by FrankBuss** [comment]
I don't think this is a good idea. The performance decrease would be minimal if always initialized with 0, as you tested yourself. I would like to have all variables initialized to default values, even float, but this is not possible with C++, unlike for example in C#, where it is impossible to use non-initialized variables. But at least you can enforce it for your own types and reducing bugs in 3rd party plugins. If the developer is not aware of the intricacies what `Vector() = default;` means (if he reads the Vector header file at all) and just assumes that an object is always initialized, like they are used to with std::string etc., it will cause more bugs in future. Not to mention all those people who just forget to initialize their variables :-)

And I guess the compiler doesn't warn that you are using uninitialized data, like it does if you do this with a simple float? At least I didn't get a warning for your ADSR.cpp.

**Comment by AndrewBelt** [comment]
The idea is to behave as closely as possible to scalar types. That trumps all other reasons.

**Comment by FrankBuss** [comment]
Ok, but maybe add a note to the plugin development tutorial. There are even people who are trying to use C++ for the first time to develop a plugin, and who have JavaScript or C# background and don't expect this behavior. And don't forget that even experienced C++ developers could do it wrong, as your ADSR.cpp demonstrated.

**Comment by ptytb** [comment]
Hey guys, I found that Fundamental ADSR misbehaves (some notes just disappear while playing) in poly mode when channels > 4 and I stumbled upon this issue, I feel it's related.

How to reproduce: start connecting a positive gate sequentially from any channel <= 4, one by one, while looking at 5th, 9th, 13th output chans of the ADSR.

![vcv_fundam_ADSR_poly](https://user-images.githubusercontent.com/2805896/93705311-e229b500-fb24-11ea-8d9a-c4aee525f6bb.png)

How to fix temporarily: always max out channels up to 16. But the catch is this will max out cpu usage as well, because 16 chans will propagate to the rest of the patch.

I'm on win10, Rack 1.1.6.
